packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -270,17 +270,179 @@ class _UntilTextBoundary extends TextBoundary {
+    // TextAlign.justify will behave exactly the same as TextAlign.start.
+    //
+    // An exception to this is when the text is not left-aligned, and the input
+    // width is double.infinity. Since the resulting Paragraph will have a width
+    // of double.infinity, and to make the text visible the paintOffset.dx is
+    // bound to be double.negativeInfinity, which invalidates all arithmetic
+    // operations.
+    final double newContentWidth = _contentWidthFor(minWidth, maxWidth, widthBasis, layout);
+    if (newContentWidth == contentWidth) {
+      return true;
+    }
+    assert(minWidth <= maxWidth);
+    // Always needsLayout when the current paintOffset and the paragraph width are not finite.
+    if (!paintOffset.dx.isFinite && !paragraph.width.isFinite && minWidth.isFinite) {
+      assert(paintOffset.dx == double.infinity);
+      assert(paragraph.width == double.infinity);
+      return false;
+    }
+    final double maxIntrinsicWidth = layout._paragraph.maxIntrinsicWidth;
+    if ((layout._paragraph.width - maxIntrinsicWidth) > -precisionErrorTolerance && (maxWidth - maxIntrinsicWidth) > -precisionErrorTolerance) {
+      // Adjust the paintOffset and contentWidth to the new input constraints.
+      contentWidth = newContentWidth;
+      return true;
+    }
+    return false;
+  }
+
+  // ---- Cached Values ----
+
+  List<TextBox>? _inlinePlaceholderBoxes;
+  List<ui.LineMetrics>? _lineMetricsCache;
+  // Holds the TextPosition the last caret metrics were computed with. When new
+  // values are passed in, we recompute the caret metrics only as necessary.
+  TextPosition? _previousCaretPosition;
+}
+
 /// This is used to cache and pass the computed metrics regarding the
 /// caret's size and position. This is preferred due to the expensive
 /// nature of the calculation.
-class _CaretMetrics {
-  const _CaretMetrics({required this.offset, this.fullHeight});
+// This should be a sealed class: A _CaretMetrics is either a _LineCaretMetrics
+// or an _EmptyLineCaretMetrics.
+class _CaretMetrics { }
+
+// The _CaretMetrics for carets located in a non-empty line. Carets located in a
+// non-empty line are associated with a glyph within the same line.
+class _LineCaretMetrics implements _CaretMetrics {
+  const _LineCaretMetrics({required this.offset, required this.writingDirection, required this.fullHeight});
   /// The offset of the top left corner of the caret from the top left
   /// corner of the paragraph.
   final Offset offset;
-
+  /// The writing direction of the glyph the _CaretMetrics is associated with.
+  final TextDirection writingDirection;
   /// The full height of the glyph at the caret position.
-  final double? fullHeight;
+  final double fullHeight;
+}
+
+// The _CaretMetrics for carets located in an empty line (when the text is
+// empty, or the caret is between two a newline characters).
+class _EmptyLineCaretMetrics implements _CaretMetrics {
```

Comment:
Seems a bit complicated with `_CaretMetrics` / `_LineCaretMetrics` `_EmptyLineCaretMetrics ` to hold a few numbers.

Also see my comments in `getFullHeightForCaret()`: I think empty lines (between newlines) should yield valid caret height since the font size is known there - the rects from `ui.Paragraph.getTextBoxesForSelection` will be zero width but will have correct height AFAIR.

So perhaps rather than assuming e.g.:
```
      return prevCodeUnit == NEWLINE_CODE_UNIT
        ? _EmptyLineCaretMetrics(lineVerticalOffset: box.bottom)
        : _LineCaretMetrics(offset: Offset(box.end, box.top), writingDirection: box.direction, fullHeight: box.bottom - box.top);
```
...we should handle any output from ui.Paragraph in `_computeCaretMetrics ()` that is:
- null rects
- zero-width rects
- regular rects
This would also help with any discrepancies between what we assume a newline is here and the engine's text layout.

Reply:
This PR doesn't change how empty lines are handled, I think the caret code changes are for caching the caret metrics differently (so "caretPrototyp" is no longer a part of the cache key). I'm working on [adding more APIs](https://docs.google.com/document/d/12iEpfcqnHMIV6slNPrbaEkWih5cuSkH-si_OTiZ9RLw/edit?resourcekey=0-GiFbE7MXJzzt50Yn1yrQQg) to improve the caret stuff. 

>  I think empty lines (between newlines) should yield valid caret height since the font size is known there

I'm planning to use the line height.

Reply:
Do we have an issue on file for this? Maybe link it here in a TODO?